### PR TITLE
Move `Script` and `ScriptBuf` to `primitives`

### PR DESCRIPTION
**Pull request by tcharding**
EDIT: See #3155 before considering looking at this.

My latest attempt to make progress on `primitives`. Draft because there is no way to merge this.

This all of the following patches on a single branch (in order)
- Patch 1: #3139 
- Patch 2: #3140
- Patch 3: #3138 
- Patch 4: #3137
- Next 6  patches are: #3134 squashed down a bunch
- Next 6 are: Create the `ScriptBufExt`
- Next 4 are: Preparation for moving types to `primitives`
- Last WIP patch is the move

### Open questions:

- Dumps rubbish in `internals/src/lib.rs`
- The additional functions on `ScriptBuf` (`push`, `pop`, `extend_from_slice`)
- How the f*** do we make this reviewable/mergable

### CODE REVIEW

**Comment by tcharding** [comment]
I believe the first 20 patches could be reviewed and merged if correct. Please review @apoelstra even though this is a draft PR.

File: internals/src/lib.rs [hunk 1]
```diff
@@ -35,3 +35,39 @@ mod parse;
 #[cfg(feature = "serde")]
 #[macro_use]
 pub mod serde;
+
+/// Reads a `usize` from an iterator of bytes.
+// TODO: Where should this live?
+pub fn read_uint_iter(
```
Comment (Kixunil) [inline internals/src/lib.rs:41]: What's this about? Why it can't be in `bitcoin`?

Comment (tcharding) [reply internals/src/lib.rs:41]: It's used in `primitives`, in this PR.

**Comment by tcharding** [comment]
I realize I may be being confusing with all these similar PRs, from @Kixunil comment [here](https://github.com/rust-bitcoin/rust-bitcoin/pull/3134#issuecomment-2275170133) I think he is going to look at #3134 not this on. Remember this PR is a super set of 3134

**Comment by tcharding** [comment]
Will rebase this one on top of #3155

**Comment by tcharding** [comment]
Replaced by #3194

**Pull request closed without merging**